src/migrate.c

```diff
@@ -0,0 +1,184 @@
+        if (req->r.migrate_keys.keys_serialized[i] == NULL) {
+            continue;
+        }
+
+        size_t key_len;
+        const char * key = RedisModule_StringPtrLen(req->r.migrate_keys.keys[i], &key_len);
+        argv[3 + (i*2)] = RedisModule_Strdup(key);
+        argv_len[3+ (i*2)] = key_len;
+
+        size_t str_len;
+        const char * str = RedisModule_StringPtrLen(req->r.migrate_keys.keys_serialized[i], &str_len);
+        argv[3 + (i*2) + 1] = RedisModule_Alloc(str_len);
+        memcpy(argv[3 + (i*2) + 1], str, str_len);
+        argv_len[3 + (i*2) + 1] = str_len;
+    }
+
+    if (redisAsyncCommandArgv(ConnGetRedisCtx(conn), transferKeysResponse, conn, argc, (const char **) argv, argv_len) != REDIS_OK) {
+        redisAsyncDisconnect(ConnGetRedisCtx(conn));
+        ConnMarkDisconnected(conn);
+    }
+
+    for(int i = 0; i < argc; i++) {
+        RedisModule_Free(argv[i]);
+    }
+    RedisModule_Free(argv);
+    RedisModule_Free(argv_len);
+}
+
+
+void MigrateKeys(RedisRaftCtx *rr, RaftReq *req)
+{
+    JoinLinkState *state = RedisModule_Calloc(1, sizeof(*state));
+    state->type = "migrate";
+    state->connect_callback = transferKeys;
+    time(&(state->start));
+    ShardGroup * sg = getShardGroupById(rr, req->r.migrate_keys.shardGroupId);
+    if (sg == NULL) {
+        RedisModule_ReplyWithError(req->ctx, "ERR couldn't resolve shardgroup id");
+        goto exit;
+    }
+    req->r.migrate_keys.migrate_term = raft_get_current_term(rr->raft);
+
+    for (size_t i = 0; i < req->r.migrate_keys.num_keys; i++) {
+        RedisModuleString *key = req->r.migrate_keys.keys[i];
+
+        if (RedisModule_KeyExists(req->ctx, key)) {
+            req->r.migrate_keys.num_serialized_keys++;
+
+            enterRedisModuleCall();
+            RedisModuleCallReply *reply = RedisModule_Call(rr->ctx, "DUMP", "s", key);
+            exitRedisModuleCall();
+
+            if (reply && RedisModule_CallReplyType(reply) == REDISMODULE_REPLY_STRING) {
+                req->r.migrate_keys.keys_serialized[i] = RedisModule_CreateStringFromCallReply(reply);
+            } else {
+                if (reply) {
+                    LOG_WARNING("unexpected response type = %d", RedisModule_CallReplyType(reply));
+                    RedisModule_FreeCallReply(reply);
+                } else {
+                    LOG_WARNING("didn't get a reply!");
+                }
+                RedisModule_ReplyWithError(req->ctx, "ERR see logs");
+                goto exit;
+            }
+        }
+    }
+
+    /* nothing to migrate, return quickly */
+    if (req->r.migrate_keys.num_serialized_keys == 0) {
+        RedisModule_ReplyWithCString(req->ctx, "OK");
+        goto exit;
+    }
+
+     for (unsigned int i = 0; i < sg->nodes_num; i++) {
+        LOG_WARNING("MigrateKeys: adding %s:%d", sg->nodes[i].addr.host, sg->nodes[i].addr.port);
+        NodeAddrListAddElement(&state->addr, &sg->nodes[i].addr);
+    }
+    state->req = req;
+
```

Comment:
really should depend on #348 here to enable proper username/password being passed to this connection